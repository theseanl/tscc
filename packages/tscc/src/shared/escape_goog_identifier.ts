/**
 * @fileoverview A valid goog.module name must start with [a-zA-Z_$] end only contain [a-zA-Z0-9._$].
 * This file provides an analogue of Javascript escape/unescape function pair for string identifiers
 * for goog.module, goog.provide, etc.
 * One does not lose information after escaping so that we can faithfully map converted module names
 * to the original TS source file's name.
 */
import path = require('path');

function codePoint(char: string) { return char.codePointAt(0); }
/**************************************************************************************************/
const LOWERCASE_A_CODE_POINT = codePoint('a')!;
const LOWERCASE_Z_CODE_POINT = codePoint('z')!;
const UPPERCASE_A_CODE_POINT = codePoint('A')!;
const UPPERCASE_Z_CODE_POINT = codePoint('Z')!;
const PERIOD_CODE_POINT = codePoint('.')!;
const LOWER_DASH_CODE_POINT = codePoint('_')!;
const DOLLAR_SIGN_CODE_POINT = codePoint('$')!;
const ZERO_CODE_POINT = codePoint('0')!;
const NINE_CODE_POINT = codePoint('9')!;
const SEP = path.sep;
/**************************************************************************************************/
function isLatin(code: number) {
	return (
		(LOWERCASE_A_CODE_POINT <= code && code <= LOWERCASE_Z_CODE_POINT) ||
		(UPPERCASE_A_CODE_POINT <= code && code <= UPPERCASE_Z_CODE_POINT));
}
function isNumber(code: number) {
	return ZERO_CODE_POINT <= code && code <= NINE_CODE_POINT;
}
function isLowerDash(code: number) {
	return code === LOWER_DASH_CODE_POINT;
}
function isPeriod(code: number) {
	return code === PERIOD_CODE_POINT;
}
function isDollarSign(code: number) {
	return code === DOLLAR_SIGN_CODE_POINT;
}
/**************************************************************************************************/
/**
 *                   Latin  ⟹  Latin
 *                  number  ⟹  number
 *                     "_"  ⟹  "_"
 *          path separator  ⟹  "." (for ergonomical reason)
 *                     "."  ⟹  "$."
 *     Any other character  ⟹  "$" followed by length 4 base36 representation of its code point,
 *                              left-padded with 0.
 *
 * This requires that the first character is not a path separator, in order to make sure that
 * the resulting escaped name does not start with ".", which is disallowed in goog.module. One should
 * always feed relative paths.
 */
export function escapeGoogAdmissibleName(name: string): string {
	let out = "";

	if (name[0] === SEP)
		throw new TypeError("Name cannot start with a path separator");

	else if (name.length > 0) {
		for (const char of name) {
			let code = codePoint(char)!;
			if (isLatin(code) || isNumber(code) || isLowerDash(code)) {
				out += char;
			} else if (char === SEP) {
				out += ".";
			} else if (isPeriod(code)) {
				out += "$.";
			} else {
				out += "$" + code.toString(36).padStart(4, "0");
			}
		}
	}

	return out;
}
export function unescapeGoogAdmissibleName(escapedName: string): string {
	let out = "";
	let i = 0;
	let code: number;
	// charCodeAt returns NaN when an index is out of range.
	while (!isNaN(code = escapedName.charCodeAt(i))) {
		if (isLatin(code) || isNumber(code) || isLowerDash(code)) {
			out += escapedName[i];
			i++;
		} else if (isPeriod(code)) {
			out += SEP;
			i++;
		} else if (isDollarSign(code)) {
			// If the next character is ".", add "."
			if (isPeriod(escapedName.charCodeAt(i + 1))) {
				out += ".";
				i += 2;
			} else {
				// Read next 4 chars
				try {
					let base32Codes = parseInt(escapedName.substr(i + 1, 4), 36);
					out += String.fromCodePoint(base32Codes);
					i += 5;
				} catch (e) {
					console.log(escapedName);
					throw new RangeError(`Invalid characters between position ${i + 1} and ${i + 4}`);
				}
			}
		} else {
			throw new RangeError(`Invalid character at position ${i}`);
		}
	}
	return out;
}
export function escapedGoogNameIsDts(escapedName: string) {
	return escapedName.endsWith("$.d$.ts");
}
