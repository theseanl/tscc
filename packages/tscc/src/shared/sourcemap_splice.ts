import { SourceMapConsumer, SourceMapGenerator, Mapping, RawSourceMap } from 'source-map';

/**
 * From a file with sourcemap, splice intervals specified with the third argument
 * and translate sourcemaps accordingly.
 * @param content
 * @param map
 * @param spliceIntervals sorted, non-overlapping intervals to splice.
 */
export default async function spliceSourceMap(content: string, map: RawSourceMap, spliceIntervals: [number, number][]): Promise<RawSourceMap> {
	const consumer = await new SourceMapConsumer(map);
	const generator = new SourceMapGenerator({ file: map.file });

	const seeker = new Seeker(content, spliceIntervals);

	consumer.eachMapping(({ source, generatedLine, generatedColumn, originalLine, originalColumn, name }) => {
		// line numbers in mozilla/source-map are 1-based. column numbers are 0-based.
		seeker.seek(generatedLine - 1, generatedColumn);
		if (seeker.isInInterval()) return;
		let transformedLine = seeker.getTransformedLine();
		let transformedColumn = seeker.getTransformedColumn();
		const mapping = getMapping(
			source, 
			transformedLine + 1, 
			transformedColumn,
			originalLine, 
			originalColumn,
			name
		);
		generator.addMapping(mapping);
	});

	return generator.toJSON();
}

/**
 * A lot of seemingly unused code was removed here that did not affect test
 * runs. If some mysterious source map bug emerges, look here first.
 */

function getMapping(
	source: string,
	generatedLine: number,
	generatedColumn: number,
	originalLine: number,
	originalColumn: number,
	name: string
): Mapping {
	return <Mapping>{
		name,
		source,
		generated: {
			line: generatedLine,
			column: generatedColumn,
		},
		original: {
			line: originalLine,
			column: originalColumn,
		},
	};
}

export function splitWithRegex(contents: string, regex: RegExp) {
	const intervals: [number, number][] = [];
	let prevEnd = 0;
	let replacedContent = '';
	let execRes: RegExpExecArray | null;
	while ((execRes = regex.exec(contents)) !== null) {
		let removeStart = execRes.index;
		let removeEnd = removeStart + execRes[0].length;
		replacedContent += contents.substring(prevEnd, removeStart);
		prevEnd = removeEnd;
		intervals.push([removeStart, removeEnd])
	}
	replacedContent += contents.substring(prevEnd);
	return { contents: replacedContent, intervals }
}

export class Seeker {
	constructor(
		private contents: string,
		private intervals: [number, number][]
	) { }
	/*************** State machine state *****************/
	// Current cursor position descriptors
	private line = 0;
	private column = 0;
	private index = 0;
	private lineStart = 0; // Index of current line's start
	private intervalIndex = -1; // Index of last interval whose start comes before then or at the same point with the current index
	private accLine = 0;
	private accColumn = 0;
	/************* State machine state end ***************/

	/**
	 * Seeks the last interval that intersects with the interval [0, index], starting from the current interval.
	 * Returns a contribution of lengths occupied by intervals in [this.Index, index).
	 * (Beware the parentheses)
	 */
	private seekInterval(index: number) {
		let { intervalIndex, index: prevIndex } = this;
		let occupied = 0;
		if (intervalIndex === -1) intervalIndex = 0;
		let interval = this.getInterval(intervalIndex);
		while (interval && interval[0] <= index) {
			if (interval[1] > prevIndex) {
				occupied += min(interval[1], index) - max(interval[0], prevIndex);
			}
			interval = this.getInterval(++intervalIndex);
		}
		this.intervalIndex = intervalIndex - 1;
		return occupied;
	}
	private seekWithinLine(nextColumn: number) {
		let increment = nextColumn - this.column;
		let nextIndex = this.index + increment;

		// Update column (line and lineStart stays the same)
		let lineEnd = this.nextLineBreak();
		if (lineEnd !== -1 && nextIndex > lineEnd)
			throw new Error('EOL');
		if (nextIndex >= this.contents.length) throw new Error('EOF');

		this.column = nextColumn;

		// Update intervalIndex, accColumn (accLine stays the same)
		let occupied = this.seekInterval(nextIndex);
		this.accColumn += increment - occupied;
		this.index = nextIndex;
	}
	private nextLine() {
		// Update line,column,lineStart
		let lineStart = this.nextLineBreak() + 1;
		if (lineStart > 0 && this.contents.length <= lineStart)
			throw new Error(`EOF`);

		this.lineStart = lineStart;
		this.line++;
		this.column = 0;
		let increment = lineStart - this.index;

		// Update interavlIndex, accLine, accColumn
		// Check if there is an interval containing lineStart - 1 (index of '\n')
		let occupied = this.seekInterval(lineStart - 1);
		let interval = this.getInterval(this.intervalIndex);

		this.index = lineStart; // Setting it after seekInterval call, as it requires prev index.

		if (!interval || interval[1] <= lineStart - 1) {
			// No interval contains the previous line break character - accLine will be increased.
			this.accLine++;
			this.accColumn = 0;
			// intervalIndex may need to proceed one step further,
			// as we are looking for the latest interval s.t. interval[0] <= lineStart
			// See test "when interval ends with \n".
			let nextInterval = this.getInterval(this.intervalIndex + 1);
			if (nextInterval && nextInterval[0] === lineStart) this.intervalIndex++;
		} else {
			// This interval contains the line break character.
			// In order to get # of occupied positions before the lineStart, we need to increase
			// occupied by 1, since the line break character is occupied by the current interval.
			this.accColumn += increment - occupied - 1;
		}
	}
	seek(nextLine: number, nextColumn: number) {
		while (nextLine > this.line) {
			this.nextLine();
		}
		this.seekWithinLine(nextColumn);
	}

	// Querying methods
	isInInterval(): boolean {
		let currentInterval = this.getInterval(this.intervalIndex);
		return currentInterval && currentInterval[1] > this.index;
	}
	getTransformedLine(): number {
		return this.accLine;
	}
	getTransformedColumn(): number {
		return this.accColumn;
	}

	private nextLineBreak(): number {
		return this.contents.indexOf('\n', this.index) || this.contents.length;
	}
	private getInterval(intervalIndex: number): [number, number] {
		return this.intervals[intervalIndex];
	}
}

function max(a: number, b: number) {
	return a > b ? a : b;
}

function min(a: number, b: number) {
	return a > b ? b : a;
}
