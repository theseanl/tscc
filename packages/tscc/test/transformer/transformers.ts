///<reference types="jest" />
import * as ts from 'typescript';
import * as tsickle from 'tsickle';
import TsccSpecWithTS from '../../src/spec/TsccSpecWithTS';
import decoratorPropertyTransformer from '../../src/transformer/decorator_property_transformer';
import restPropertyTransformer from '../../src/transformer/rest_property_transformer';
import dtsRequireTypeTransformer from '../../src/transformer/dts_requiretype_transformer';
import ITsccSpecWithTS from '../../src/spec/ITsccSpecWithTS';
import {escapeGoogAdmissibleName} from '../../src/shared/escape_goog_identifier';
import path = require('path');
import upath = require('upath');

const samplesRoot = path.resolve(__dirname, '../sample');

describe(`decoratorPropertyTransformer`, function () {
	test(`modifies property name access to goog.reflect.objectProperty, add @nocollapse to jsdoc`, () => {
		const testFiles = [
			"decorator/decorates.ts"
		];

		const emitWithDecoratorTransformations = (override?: Partial<ts.CompilerOptions>) => {
			return emit(testFiles, (transformerHost) => ({
				afterTs: [decoratorPropertyTransformer(transformerHost)]
			}), override);
		}

		const {out} = emitWithDecoratorTransformations();
		for (let testFile of testFiles) {
			expect(out.get(testFile)).toMatchSnapshot(testFile);
		}

		const {out: es5Out} = emitWithDecoratorTransformations({target: ts.ScriptTarget.ES5});
		for (let testFile of testFiles) {
			expect(es5Out.get(testFile)).toMatchSnapshot(testFile + ' - es5');
		}

		const {out: es3Out} = emitWithDecoratorTransformations({target: ts.ScriptTarget.ES3});
		for (let testFile of testFiles) {
			expect(es3Out.get(testFile)).toMatchSnapshot(testFile + ' - es3');
		}
	})
})

describe(`restPropertyTransformer`, function () {
	test(`modifies property name access to goog.reflect.objectProperty`, () => {
		const testFiles = [
			"rest/case_1.ts"
		];

		const emitWithRestTransformations = (override?: Partial<ts.CompilerOptions>) => {
			return emit(testFiles, (transformerHost) => ({
				afterTs: [restPropertyTransformer(transformerHost)]
			}), override);
		};

		const {out} = emitWithRestTransformations();
		for (let testFile of testFiles) {
			expect(out.get(testFile)).toMatchSnapshot(testFile);
		}

		const {out: es5Out} = emitWithRestTransformations({target: ts.ScriptTarget.ES5});
		for (let testFile of testFiles) {
			expect(es5Out.get(testFile)).toMatchSnapshot(testFile + ' - es5');
		}

		const {out: es3Out} = emitWithRestTransformations({target: ts.ScriptTarget.ES3});
		for (let testFile of testFiles) {
			expect(es3Out.get(testFile)).toMatchSnapshot(testFile + ' - es3');
		}
	});

	test(`works well when combined with decorators`, () => {
		const testFiles = [
			"rest/combined_with_decorators.ts"
		];

		const emitWithRestTransformations = (override?: Partial<ts.CompilerOptions>) => {
			return emit(testFiles, (transformerHost) => ({
				afterTs: [
					decoratorPropertyTransformer(transformerHost),
					restPropertyTransformer(transformerHost)
				]
			}), override);
		};

		const {out} = emitWithRestTransformations();
		for (let testFile of testFiles) {
			expect(out.get(testFile)).toMatchSnapshot(testFile);
		}

		const {out: es5Out} = emitWithRestTransformations({target: ts.ScriptTarget.ES5});
		for (let testFile of testFiles) {
			expect(es5Out.get(testFile)).toMatchSnapshot(testFile + ' - es5');
		}

		const {out: es3Out} = emitWithRestTransformations({target: ts.ScriptTarget.ES3});
		for (let testFile of testFiles) {
			expect(es3Out.get(testFile)).toMatchSnapshot(testFile + ' - es3');
		}
	})
});

describe(`dts_requiretype_transformer`, () => {
	test(`modifies requireType calls to a global variable`, () => {
		const testFiles = [
			"dts_requiretype/entry.ts"
		];

		const mockSpec = <TsccSpecWithTS>{
			getTSRoot() {
				return samplesRoot;
			},
			getExternalModuleNames() {
				return []
			}
		};

		const tsickleHostOverride: Partial<tsickle.TsickleHost> = {
			googmodule: true,
			transformTypesToClosure: true,
			pathToModuleName(context, fileName) {
				// TODO: Having to depend on actual escapeGoog function here looks bad.
				// We have to decouple this test from an actual implementation. Maybe introduce a
				// tsccHost object that contains escapeGoog** and instantiate tsickleHost using that.
				// Currently
				if (/tslib/.test(fileName)) return 'tslib';
				if (context.length > 0) fileName = path.resolve(path.dirname(context), fileName);
				const rel = path.relative(samplesRoot, fileName);
				return escapeGoogAdmissibleName(rel);
			},
			fileNameToModuleId: x => ""
		};

		const {out} = emit(testFiles, (transformerHost) => ({
			afterTs: [dtsRequireTypeTransformer(mockSpec, transformerHost)]
		}), {}, tsickleHostOverride);

		for (let testFile of testFiles) 
			expect(out.get(testFile)).toMatchSnapshot(testFile);
	})
});

type EmitTransformerFactory = (host: tsickle.TsickleHost) => tsickle.EmitTransformers;

function emit(
	files: string[], transformerFactory: EmitTransformerFactory,
	override: Partial<ts.CompilerOptions> = {}, tsickleHostOverride: Partial<tsickle.TsickleHost> = {}
) {
	const {parsedConfig} = TsccSpecWithTS.loadTsConfigFromPath(samplesRoot);
	const {options} = parsedConfig;
	Object.assign(options, override);
	const host = ts.createCompilerHost(options);
	files = files.map(file => path.resolve(samplesRoot, file));
	const program = ts.createProgram(files, options, host);
	const transformerHost = Object.assign({
		shouldSkipTsickleProcessing: () => false,
		shouldIgnoreWarningsForPath: () => true,
		googmodule: false,
		pathToModuleName: x => x,
		fileNameToModuleId: x => x,
		moduleResolutionHost: ts.createCompilerHost(parsedConfig.options),
		options,
	}, tsickleHostOverride);

	const out = new Map<string, string>();
	const {externs} = tsickle.emitWithTsickle(
		program, transformerHost, host, options, undefined,
		(fileName, data) => {
			fileName = path.relative(samplesRoot, fileName);
			fileName = fileName.slice(0, -path.extname(fileName).length) + '.ts';
			fileName = upath.toUnix(fileName); // so that test outputs are independent of platforms
			out.set(fileName, data);
		},
		undefined /* cancellationtoken */, false /* emitOnlyDtsFiles */,
		transformerFactory(transformerHost)
	);
	return {out, externs}
}
