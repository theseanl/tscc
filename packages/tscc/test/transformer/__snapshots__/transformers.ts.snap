// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decoratorPropertyTransformer modifies property name access to goog.reflect.objectProperty, add @nocollapse to jsdoc: decorator/decorates.ts - es3 1`] = `
"const tscc_goog_reflect_injected = goog.require(\\"goog.reflect\\");
\\"use strict\\";
exports.__esModule = true;
var tslib_1 = require(\\"tslib\\");
var adornClass = function (target) { return target; };
var adornProperty = function (target, prop) { };
var adornMethod = function (target, prop, desc) { return desc; };
var adornParameter = function (target, prop, paramIndex) { };
var A = /** @class */ (function () {
    function A() {
    }
    A.prototype.aMethod = function () {
    };
    Object.defineProperty(A.prototype, \\"anAccessor\\", {
        get: function () {
            return NaN;
        },
        set: function (v) {
        },
        enumerable: false,
        configurable: true
    });
    A.prototype.aMethodWithParams = function (param) {
    };
    tslib_1.__decorate([
        adornMethod
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aMethod\\", A.prototype));
    tslib_1.__decorate([
        adornProperty
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aProperty\\", A.prototype));
    tslib_1.__decorate([
        adornMethod
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"anAccessor\\", A.prototype));
    tslib_1.__decorate([
        tslib_1.__param(0, adornParameter)
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aMethodWithParams\\", A.prototype));
    A = tslib_1.__decorate([
        adornClass
    ], A);
    return A;
}());
\\"__tscc_export_start__\\";
{
    self[\\"tscc_global_access_name_1\\"] = (A.prototype).aMethod;
    self[\\"tscc_global_access_name_2\\"] = (A.prototype).aProperty;
    self[\\"tscc_global_access_name_3\\"] = (A.prototype).anAccessor;
    self[\\"tscc_global_access_name_4\\"] = (A.prototype).aMethodWithParams;
}
\\"__tscc_export_end__\\";
"
`;

exports[`decoratorPropertyTransformer modifies property name access to goog.reflect.objectProperty, add @nocollapse to jsdoc: decorator/decorates.ts - es5 1`] = `
"const tscc_goog_reflect_injected = goog.require(\\"goog.reflect\\");
\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var tslib_1 = require(\\"tslib\\");
var adornClass = function (target) { return target; };
var adornProperty = function (target, prop) { };
var adornMethod = function (target, prop, desc) { return desc; };
var adornParameter = function (target, prop, paramIndex) { };
var A = /** @class */ (function () {
    function A() {
    }
    A.prototype.aMethod = function () {
    };
    Object.defineProperty(A.prototype, \\"anAccessor\\", {
        get: function () {
            return NaN;
        },
        set: function (v) {
        },
        enumerable: false,
        configurable: true
    });
    A.prototype.aMethodWithParams = function (param) {
    };
    tslib_1.__decorate([
        adornMethod
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aMethod\\", A.prototype), null);
    tslib_1.__decorate([
        adornProperty
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aProperty\\", A.prototype), void 0);
    tslib_1.__decorate([
        adornMethod
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"anAccessor\\", A.prototype), null);
    tslib_1.__decorate([
        tslib_1.__param(0, adornParameter)
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aMethodWithParams\\", A.prototype), null);
    A = tslib_1.__decorate([
        adornClass
    ], A);
    return A;
}());
\\"__tscc_export_start__\\";
{
    self[\\"tscc_global_access_name_1\\"] = (A.prototype).aMethod;
    self[\\"tscc_global_access_name_2\\"] = (A.prototype).aProperty;
    self[\\"tscc_global_access_name_3\\"] = (A.prototype).anAccessor;
    self[\\"tscc_global_access_name_4\\"] = (A.prototype).aMethodWithParams;
}
\\"__tscc_export_end__\\";
"
`;

exports[`decoratorPropertyTransformer modifies property name access to goog.reflect.objectProperty, add @nocollapse to jsdoc: decorator/decorates.ts 1`] = `
"const tscc_goog_reflect_injected = goog.require(\\"goog.reflect\\");
\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const tslib_1 = require(\\"tslib\\");
const adornClass = (target) => target;
const adornProperty = (target, prop) => { };
const adornMethod = (target, prop, desc) => desc;
const adornParameter = (target, prop, paramIndex) => { };
let A = class A {
    aMethod() {
    }
    get anAccessor() {
        return NaN;
    }
    set anAccessor(v) {
    }
    aMethodWithParams(param) {
    }
};
tslib_1.__decorate([
    adornMethod
], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aMethod\\", A.prototype), null);
tslib_1.__decorate([
    adornProperty
], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aProperty\\", A.prototype), void 0);
tslib_1.__decorate([
    adornMethod
], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"anAccessor\\", A.prototype), null);
tslib_1.__decorate([
    tslib_1.__param(0, adornParameter)
], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"aMethodWithParams\\", A.prototype), null);
A = tslib_1.__decorate([
    adornClass
], A);
\\"__tscc_export_start__\\";
{
    self[\\"tscc_global_access_name_1\\"] = (A.prototype).aMethod;
    self[\\"tscc_global_access_name_2\\"] = (A.prototype).aProperty;
    self[\\"tscc_global_access_name_3\\"] = (A.prototype).anAccessor;
    self[\\"tscc_global_access_name_4\\"] = (A.prototype).aMethodWithParams;
}
\\"__tscc_export_end__\\";
"
`;

exports[`dts_requiretype_transformer modifies requireType calls to a global variable: dts_requiretype/entry.ts 1`] = `
"/**
 * @fileoverview added by tsickle
 * Generated from: dts_requiretype/entry.ts
 * @suppress {checkTypes,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
goog.module('dts_requiretype.entry$.ts');
var module = module || { id: '' };
goog.require('tslib');
const tsickle_required_1 = dts_requiretype$required$$d$$ts;
const required = goog.require('dts_requiretype.required');
/** @type {!tsickle_required_1.I} */
var a = required.func();
"
`;

exports[`restPropertyTransformer modifies property name access to goog.reflect.objectProperty: rest/case_1.ts - es3 1`] = `
"const tscc_goog_reflect_injected = goog.require(\\"goog.reflect\\");
\\"use strict\\";
var _a;
exports.__esModule = true;
var tslib_1 = require(\\"tslib\\");
var o = (_a = {
        a: {},
        b: {},
        \\"c\\": {}
    },
    _a[\\"d\\"] = {},
    _a[Symbol('foo')] = {},
    _a[Math.random() > .5 ? \\"foo\\" : \\"bar\\"] = {},
    _a);
var a = o.a, c = o.c, d = o.d, foo = o.foo, rest = tslib_1.__rest(o, [tscc_goog_reflect_injected.objectProperty(\\"a\\", o), tscc_goog_reflect_injected.objectProperty(\\"c\\", o), tscc_goog_reflect_injected.objectProperty(\\"d\\", o), tscc_goog_reflect_injected.objectProperty(\\"foo\\", o)]);
function A(a, _a) {
    var b = _a.b, c = _a.c, d = tslib_1.__rest(_a, [tscc_goog_reflect_injected.objectProperty(\\"b\\", _a), tscc_goog_reflect_injected.objectProperty(\\"c\\", _a)]);
    console.log(d);
}
// From what's new in Typescript 2.1
(function () {
    var original = { a: {}, b: {}, c: {} };
    var copy = tslib_1.__assign({}, original);
})();
(function () {
    var foo = {}, bar = {}, baz = {};
    var merged = tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, foo), bar), baz);
})();
(function () {
    var obj = { x: 1, y: \\"string\\" };
    var newObj = tslib_1.__assign(tslib_1.__assign({}, obj), { z: 3, y: 4 }); // { x: number, y: number, z: number }
})();
(function () {
    var obj = { x: 1, y: 1, z: 1 };
    var z = obj.z, obj1 = tslib_1.__rest(obj, [tscc_goog_reflect_injected.objectProperty(\\"z\\", obj)]);
    obj1; // {x: number, y:number};
})();
"
`;

exports[`restPropertyTransformer modifies property name access to goog.reflect.objectProperty: rest/case_1.ts - es5 1`] = `
"const tscc_goog_reflect_injected = goog.require(\\"goog.reflect\\");
\\"use strict\\";
var _a;
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var tslib_1 = require(\\"tslib\\");
var o = (_a = {
        a: {},
        b: {},
        \\"c\\": {}
    },
    _a[\\"d\\"] = {},
    _a[Symbol('foo')] = {},
    _a[Math.random() > .5 ? \\"foo\\" : \\"bar\\"] = {},
    _a);
var a = o.a, c = o.c, d = o.d, foo = o.foo, rest = tslib_1.__rest(o, [tscc_goog_reflect_injected.objectProperty(\\"a\\", o), tscc_goog_reflect_injected.objectProperty(\\"c\\", o), tscc_goog_reflect_injected.objectProperty(\\"d\\", o), tscc_goog_reflect_injected.objectProperty(\\"foo\\", o)]);
function A(a, _a) {
    var b = _a.b, c = _a.c, d = tslib_1.__rest(_a, [tscc_goog_reflect_injected.objectProperty(\\"b\\", _a), tscc_goog_reflect_injected.objectProperty(\\"c\\", _a)]);
    console.log(d);
}
// From what's new in Typescript 2.1
(function () {
    var original = { a: {}, b: {}, c: {} };
    var copy = tslib_1.__assign({}, original);
})();
(function () {
    var foo = {}, bar = {}, baz = {};
    var merged = tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, foo), bar), baz);
})();
(function () {
    var obj = { x: 1, y: \\"string\\" };
    var newObj = tslib_1.__assign(tslib_1.__assign({}, obj), { z: 3, y: 4 }); // { x: number, y: number, z: number }
})();
(function () {
    var obj = { x: 1, y: 1, z: 1 };
    var z = obj.z, obj1 = tslib_1.__rest(obj, [tscc_goog_reflect_injected.objectProperty(\\"z\\", obj)]);
    obj1; // {x: number, y:number};
})();
"
`;

exports[`restPropertyTransformer modifies property name access to goog.reflect.objectProperty: rest/case_1.ts 1`] = `
"const tscc_goog_reflect_injected = goog.require(\\"goog.reflect\\");
\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const tslib_1 = require(\\"tslib\\");
var o = {
    a: {},
    b: {},
    \\"c\\": {},
    [\\"d\\"]: {},
    [Symbol('foo')]: {},
    [Math.random() > .5 ? \\"foo\\" : \\"bar\\"]: {}
};
var { a, c, d, foo } = o, rest = tslib_1.__rest(o, [tscc_goog_reflect_injected.objectProperty(\\"a\\", o), tscc_goog_reflect_injected.objectProperty(\\"c\\", o), tscc_goog_reflect_injected.objectProperty(\\"d\\", o), tscc_goog_reflect_injected.objectProperty(\\"foo\\", o)]);
function A(a, _a) {
    var { b, c } = _a, d = tslib_1.__rest(_a, [tscc_goog_reflect_injected.objectProperty(\\"b\\", _a), tscc_goog_reflect_injected.objectProperty(\\"c\\", _a)]);
    console.log(d);
}
// From what's new in Typescript 2.1
(function () {
    var original = { a: {}, b: {}, c: {} };
    let copy = Object.assign({}, original);
})();
(function () {
    var foo = {}, bar = {}, baz = {};
    let merged = Object.assign(Object.assign(Object.assign({}, foo), bar), baz);
})();
(function () {
    let obj = { x: 1, y: \\"string\\" };
    var newObj = Object.assign(Object.assign({}, obj), { z: 3, y: 4 }); // { x: number, y: number, z: number }
})();
(function () {
    let obj = { x: 1, y: 1, z: 1 };
    let { z } = obj, obj1 = tslib_1.__rest(obj, [tscc_goog_reflect_injected.objectProperty(\\"z\\", obj)]);
    obj1; // {x: number, y:number};
})();
"
`;

exports[`restPropertyTransformer works well when combined with decorators: rest/combined_with_decorators.ts - es3 1`] = `
"const tscc_goog_reflect_injected = goog.require(\\"goog.reflect\\");
\\"use strict\\";
exports.__esModule = true;
var tslib_1 = require(\\"tslib\\");
function myDecorator(target, prop, desc) {
    var value = desc.value, rest = tslib_1.__rest(desc, [tscc_goog_reflect_injected.objectProperty(\\"value\\", desc)]);
    console.log(rest);
    desc.value = {};
}
var A = /** @class */ (function () {
    function A() {
    }
    A.prototype.myMethod = function () { };
    tslib_1.__decorate([
        myDecorator
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"myMethod\\", A.prototype));
    return A;
}());
var b = new A();
window.a = b.myMethod();
var _a = { a: {}, b: {}, c: {} }, c = _a.c, rest = tslib_1.__rest(_a, [tscc_goog_reflect_injected.objectProperty(\\"c\\", _a)]);
window.d = rest;
\\"__tscc_export_start__\\";
{
    self[\\"tscc_global_access_name_1\\"] = (A.prototype).myMethod;
}
\\"__tscc_export_end__\\";
"
`;

exports[`restPropertyTransformer works well when combined with decorators: rest/combined_with_decorators.ts - es5 1`] = `
"const tscc_goog_reflect_injected = goog.require(\\"goog.reflect\\");
\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var tslib_1 = require(\\"tslib\\");
function myDecorator(target, prop, desc) {
    var value = desc.value, rest = tslib_1.__rest(desc, [tscc_goog_reflect_injected.objectProperty(\\"value\\", desc)]);
    console.log(rest);
    desc.value = {};
}
var A = /** @class */ (function () {
    function A() {
    }
    A.prototype.myMethod = function () { };
    tslib_1.__decorate([
        myDecorator
    ], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"myMethod\\", A.prototype), null);
    return A;
}());
var b = new A();
window.a = b.myMethod();
var _a = { a: {}, b: {}, c: {} }, c = _a.c, rest = tslib_1.__rest(_a, [tscc_goog_reflect_injected.objectProperty(\\"c\\", _a)]);
window.d = rest;
\\"__tscc_export_start__\\";
{
    self[\\"tscc_global_access_name_1\\"] = (A.prototype).myMethod;
}
\\"__tscc_export_end__\\";
"
`;

exports[`restPropertyTransformer works well when combined with decorators: rest/combined_with_decorators.ts 1`] = `
"const tscc_goog_reflect_injected = goog.require(\\"goog.reflect\\");
\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const tslib_1 = require(\\"tslib\\");
function myDecorator(target, prop, desc) {
    let { value } = desc, rest = tslib_1.__rest(desc, [tscc_goog_reflect_injected.objectProperty(\\"value\\", desc)]);
    console.log(rest);
    desc.value = {};
}
class A {
    myMethod() { }
}
tslib_1.__decorate([
    myDecorator
], A.prototype, tscc_goog_reflect_injected.objectProperty(\\"myMethod\\", A.prototype), null);
const b = new A();
window.a = b.myMethod();
const _a = { a: {}, b: {}, c: {} }, { c } = _a, rest = tslib_1.__rest(_a, [tscc_goog_reflect_injected.objectProperty(\\"c\\", _a)]);
window.d = rest;
\\"__tscc_export_start__\\";
{
    self[\\"tscc_global_access_name_1\\"] = (A.prototype).myMethod;
}
\\"__tscc_export_end__\\";
"
`;
